### PR TITLE
[Cmake] Clean Lib dependencies

### PR DESCRIPTION
**Pull request by nonifier**
This is what the libraries dependencies between ours modules used to be: [previous_dependency_graph.pdf](https://github.com/CanalTP/navitia/files/3805049/graph.prev.pdf)

This is now what it looks like: [graph.pdf](https://github.com/CanalTP/navitia/files/3805000/graph.pdf)

### The main goals for this work:
 * Speed up development time by only rebuilding what's necessary.
 * Speed up linking time by removing unwanted libs.
 * Clearer dependency graph to help organise the code better.

Here is what Kraken depends on : [graph.kraken.pdf](https://github.com/CanalTP/navitia/files/3805091/graph.kraken.pdf). A few things could be done around `types` `utils` and `data`.

### Cmake tips of the day: 
To generate the previous graphs, I've used :
```bash
cmake --graphviz=graphviz.dot ../path/to/navitia/source
dot -Tpdf -o graph.pdf graphviz.dot
```
 
## Requirements
- [x] https://github.com/CanalTP/SimpleAmqpClient/pull/1
- [x] https://github.com/CanalTP/utils/pull/84

### CODE REVIEW

**Comment by kinnou02** [comment]
that's a lot of cmake to read :D

File: source/ed/CMakeLists.txt [hunk 1]
```diff
@@ -4,51 +4,50 @@ SET(SOURCE_LIB
...
-
+target_link_libraries(ed types fare)
 
 FIND_LIBRARY(OSMPBF osmpbf)
 
 add_library(osm2ed_lib osm2ed.cpp)
+target_link_libraries(osm2ed_lib ed transportation_data_import ${OSMPBF} protobuf z ${Boost_PROGRAM_OPTIONS_LIBRARY})
 add_executable(osm2ed osm2ed_main.cpp)
-target_link_libraries(osm2ed osm2ed_lib transportation_data_import ed connectors types ${PQXX_LIB} ${OSMPBF}
-  pb_lib utils ${BOOST_LIBS} log4cplus z protobuf ${NAVITIA_ALLOCATOR})
```
Comment (kinnou02) [inline source/ed/CMakeLists.txt:20]: I would like to keep `NAVITIA_ALLOCATOR`

File: source/routing/CMakeLists.txt [hunk 2]
```diff
@@ -9,18 +6,15 @@ SET(ROUTING_SRC
   journey.cpp)
 
 add_library(routing ${ROUTING_SRC})
-add_dependencies(routing protobuf_files)
-target_link_libraries(routing types fare georef utils autocomplete ${BOOST_LIBS})
+target_link_libraries(routing fare georef autocomplete pthread)
```
Comment (kinnou02) [inline source/routing/CMakeLists.txt:9]: It's outside the scope of this PR but this is very wrong....
`routing` shouldn't depend on `autocomplete` nor `fare`, for that we at least need to put `raptor_api` outside `routing`

File: source/tests/mock-kraken/CMakeLists.txt [hunk 3]
```diff
@@ -1,8 +1,11 @@
 include_directories("..")
 
-set(ALL_LIBS workers ed disruption_api calendar_api time_tables types autocomplete proximitylist
```
Comment (benoit-bst) [inline source/tests/mock-kraken/CMakeLists.txt:3]: hahaaa, link with the world :world_map:

**Review by benoit-bst — APPROVED** [review]
**Quote of the day** **:scroll:** 
   The link forest :evergreen_tree:  will become organize fields :sunrise_over_mountains:. That's the progress :hammer_and_wrench:

File: source/ed/CMakeLists.txt (hunk 1 above)
Comment (nonifier) [reply source/ed/CMakeLists.txt:20]: added back !

File: source/routing/CMakeLists.txt (hunk 2 above)
Comment (nonifier) [reply source/routing/CMakeLists.txt:9]: I do agree ! And I'm happy you raised that point as it was the purpose of this work :)

File: source/tests/mock-kraken/CMakeLists.txt [hunk 4]
```diff
@@ -1,8 +1,11 @@
...
-    ptreferential time_tables data routing fare types georef utils SimpleAmqpClient
-    rabbitmq-static log4cplus ${Boost_LIBRARIES} protobuf ${NAVITIA_ALLOCATOR})
+set(ALL_LIBS
```
Comment (pbougue) [inline source/tests/mock-kraken/CMakeLists.txt:3]: Shoudn't we rename it, then?

File: source/tests/chaos/CMakeLists.txt [hunk 5]
```diff
@@ -2,7 +2,12 @@
 ## Setup the BOOST tests that will exercise Chaos database
 #
 add_executable(chaos_db_tests chaos_db_tests.cpp)
-target_link_libraries(chaos_db_tests make_disruption_from_chaos apply_disruption ed workers data types pb_lib utils log4cplus ${NAVITIA_ALLOCATOR} ${Boost_LIBRARIES} ${Boost_DATE_TIME_LIBRARY} protobuf)
+target_link_libraries(chaos_db_tests
+    make_disruption_from_chaos
+    ed
+    workers
+    ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY}
```
Comment (pbougue) [inline source/tests/chaos/CMakeLists.txt:9]: Can't you link to ALL_LIBS?
(the question seems to be worthy for multiple places)

File: source/routing/tests/CMakeLists.txt [hunk 6]
```diff
@@ -1,58 +1,54 @@
-#We use the BOOST_LIBS define is the parent: routing
-SET(BOOST_LIBS ${BOOST_LIBS} ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY})
+set(RAPTOR_LINK_LIBS
+    routing
+    ed
+    workers
+    ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY}
+    ${NAVITIA_ALLOCATOR}
```
Comment (pbougue) [inline source/routing/tests/CMakeLists.txt:6]: These are ALL_LIBS

File: source/ptreferential/tests/CMakeLists.txt [hunk 7]
```diff
@@ -1,26 +1,24 @@
-
+set(PTREF_TEST_LINK_LIBS
+    ed
+    ptreferential
+    ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY}
```
Comment (pbougue) [inline source/ptreferential/tests/CMakeLists.txt:4]: Same ALL_LIBS (except alloc, but it may be nicer)

File: source/kraken/tests/CMakeLists.txt [hunk 8]
```diff
@@ -1,39 +1,41 @@
-#We use the BOOST_LIBS define is the parent
-SET(BOOST_LIBS ${BOOST_LIBS} ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY})
+set(KRAKEN_TEST_LINK_LIBS
+    workers
+    ${NAVITIA_ALLOCATOR}
+    ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY}
```
Comment (pbougue) [inline source/kraken/tests/CMakeLists.txt:4]: ALL_LIBS minus ed : maybe it's worth a variable too?

File: source/time_tables/tests/CMakeLists.txt [hunk 9]
```diff
@@ -5,21 +5,24 @@ SET(BOOST_LIBS ${Boost_UNIT_TEST_FRAMEWORK_LIBRARY} ${Boost_FILESYSTEM_LIBRARY}
...
-target_link_libraries(thermometer_test time_tables ptreferential ed data fare routing utils pb_lib thermometer georef ${BOOST_LIBS} log4cplus protobuf)
+target_link_libraries(thermometer_test thermometer ${BOOST_LIBS})
 ADD_BOOST_TEST(thermometer_test)
 
-
+set(TIME_TABLE_LINK_LIBS
+    time_tables
+    ed
+    data
+    ${BOOST_LIBS}
```
Comment (pbougue) [inline source/time_tables/tests/CMakeLists.txt:15]: Seems like too many libs are into this, filesystem, regex and iostream I'm not sure

**Review by pbougue — APPROVED** [review]
:heart_eyes: great job!
Will trigger other interesting things... :+1: 

Don't know how to enforce this coding behavior in the future, but this definitely helps for cleanness.
Maybe add the graph generation cmd to contributing doc?

**Pull request merged**